core/src/main/scala/cats/data/NonEmptySet.scala

```diff
@@ -25,15 +25,21 @@ package data
 import cats.kernel._
 
 import scala.collection.immutable._
+
 import kernel.compat.scalaVersionSpecific._
 
-private[data] object NonEmptySetImpl extends NonEmptySetInstances with Newtype {
+private[data] object NonEmptySetImpl extends NonEmptySetImpl with Newtype {
 
   private[data] def create[A](s: SortedSet[A]): Type[A] =
     s.asInstanceOf[Type[A]]
 
   private[data] def unwrap[A](s: Type[A]): SortedSet[A] =
     s.asInstanceOf[SortedSet[A]]
+}
+
+// This class has to be public to expose public methods to Mima.
+@deprecated("for internal purposes only - do not use", "2.8.0")
+sealed abstract class NonEmptySetImpl extends NonEmptySetInstances { self: NonEmptySetImpl.type =>
```

Comment:
Will this put a deprecation on any use of the methods in this class?

Reply:
Good question, I'll check it out, thanks.

Reply:
@armanbilge seems that the compiler gets worried about the `NonEmptyChainImpl` type only but not any of its methods (which is expected). Furthermore, only Scala 2.12 and 2.13 get worried, whereas Scala 3.0.2 silently ignores the deprecation at all (iiric, this issue was also discussed previously).

Reply:
I'm still not 100% sure that this change:
 - will really help Mima to track those public methods
 - will not break bin-compat with previous Cats versions

Reply:
> will really help Mima to track those public methods

I'm confident about this. If MiMa doesn't track public classes or methods (like this one) then we're in big trouble 😂 

> will not break bin-compat with previous Cats versions

This is harder to be confident of. Based on my experience I don't expect it to break, but we have no MiMa support on anything from 2.0-2.7. Our best bet is to add some of these methods to the manual bincompat test, that is the only sure way.

> seems that the compiler gets worried about the NonEmptyChainImpl type only but not any of its methods (which is expected)

Thanks for checking the deprecation thing. Somehow I'm surprised about this but great that it works! :)

Reply:
@satorg so I did an experiment in 783200628719cceae8cd47614419dac84b818d34, where I made the NonEmpty objects public. My intention was to publish a snapshot of that, which we could use for bincompat checking in this PR.

It turns out that making those objects public was sufficient to reactivate MiMa on them (actually I think this is MiMa bug, but let's ignore that for now). And furthermore, I discovered that we've already broken bincompat with Cats 1.x:
https://github.com/typelevel/cats/runs/6740199306?check_suite_focus=true#step:15:38

Reply:
Cool, with this patch bincompat is fixed.
```patch
diff --git a/core/src/main/scala/cats/data/NonEmptySet.scala b/core/src/main/scala/cats/data/NonEmptySet.scala
index 89ecd5b52..1d074ede7 100644
--- a/core/src/main/scala/cats/data/NonEmptySet.scala
+++ b/core/src/main/scala/cats/data/NonEmptySet.scala
@@ -433,7 +433,7 @@ sealed abstract private[data] class NonEmptySetInstances0 extends NonEmptySetIns
 }
 
 sealed abstract private[data] class NonEmptySetInstances1 {
-  private[data] def catsDataEqForNonEmptySet[A](implicit A: Order[A]): Eq[NonEmptySet[A]] =
+  def catsDataEqForNonEmptySet[A](implicit A: Order[A]): Eq[NonEmptySet[A]] =
     catsDataEqForNonEmptySetFromEqA[A]
 
   implicit def catsDataEqForNonEmptySetFromEqA[A](implicit A: Eq[A]): Eq[NonEmptySet[A]] =
```

Reply:
The MiMa bug I mentioned, I suspect is a variation of https://github.com/lightbend/mima/pull/699.